checker/src/org/checkerframework/checker/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,157 @@
+import org.checkerframework.dataflow.analysis.FlowExpressions;
+import org.checkerframework.dataflow.analysis.FlowExpressions.FieldAccess;
+import org.checkerframework.dataflow.analysis.FlowExpressions.LocalVariable;
+import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.framework.flow.CFAbstractAnalysis;
+import org.checkerframework.framework.flow.CFAbstractStore;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+public class UpperBoundStore extends CFAbstractStore<CFValue, UpperBoundStore> {
+
+    protected UpperBoundStore(UpperBoundStore other) {
+        super(other);
+    }
+
+    public UpperBoundStore(
+            CFAbstractAnalysis<CFValue, UpperBoundStore, ?> analysis, boolean sequentialSemantics) {
+        super(analysis, sequentialSemantics);
+    }
+
+    // If something is removed from a list it reduces the minlen of anything that could be an alias of the list by 1.
+    // If a list is cleared, anything that could be an alias of this list goes to UpperBoundUnknown.
+    @Override
+    public void updateForMethodCall(
+            MethodInvocationNode n, AnnotatedTypeFactory atypeFactory, CFValue val) {
+        Receiver caller = FlowExpressions.internalReprOf(atypeFactory, n.getTarget().getReceiver());
+        String methodName = n.getTarget().getMethod().toString();
+        boolean remove = methodName.startsWith("remove(");
+        boolean clear = methodName.startsWith("clear(");
+        boolean add = methodName.startsWith("add(");
+        Map<Receiver, CFValue> replace = new HashMap<Receiver, CFValue>();
+        if (clear) {
+            for (FlowExpressions.LocalVariable rec : localVariableValues.keySet()) {
+                applyTransfer(rec, replace, true, atypeFactory, caller);
+            }
+            for (FieldAccess rec : fieldValues.keySet()) {
+                applyTransfer(rec, replace, true, atypeFactory, caller);
+            }
+        }
+        if (remove) {
+            for (FlowExpressions.LocalVariable rec : localVariableValues.keySet()) {
+                applyTransfer(rec, replace, false, atypeFactory, caller);
+            }
+            for (FieldAccess rec : fieldValues.keySet()) {
+                applyTransfer(rec, replace, false, atypeFactory, caller);
+            }
+        }
+        if (add) {
+            for (FlowExpressions.LocalVariable rec : localVariableValues.keySet()) {
+                applyAdd(rec, replace, atypeFactory, caller);
+            }
+            for (FieldAccess rec : fieldValues.keySet()) {
+                applyAdd(rec, replace, atypeFactory, caller);
+            }
+        }
+        for (Receiver rec : replace.keySet()) {
+            replaceValue(rec, replace.get(rec));
```

Comment:
Is there reason why each method doesn't call replaceValue directly?  (If so, please add a comment a loop through the  entry set so you don't have to call Map.get.)